File: src/main/java/com/exactpro/th2/common/schema/message/impl/rabbitmq/connection/ConsumeConnectionManager.java
package com.exactpro.th2.common.schema.message.impl.rabbitmq.connection;

import java.io.IOException;
import java.util.List;
import java.util.Objects;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicInteger;

import org.jetbrains.annotations.NotNull;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.exactpro.th2.common.metrics.HealthMetrics;
import com.exactpro.th2.common.schema.message.impl.rabbitmq.configuration.ConnectionManagerConfiguration;
import com.google.common.util.concurrent.ThreadFactoryBuilder;
import com.rabbitmq.client.BlockedListener;
import com.rabbitmq.client.Channel;
import com.rabbitmq.client.Connection;
import com.rabbitmq.client.ConnectionFactory;
import com.rabbitmq.client.Consumer;
import com.rabbitmq.client.ExceptionHandler;
import com.rabbitmq.client.Recoverable;
import com.rabbitmq.client.RecoveryListener;
import com.rabbitmq.client.TopologyRecoveryException;

public class ConsumeConnectionManager {

    private static final Logger LOGGER = LoggerFactory.getLogger(ConsumeConnectionManager.class);
    private final Connection connection;
    private final AtomicInteger connectionRecoveryAttempts = new AtomicInteger(0);
    private final AtomicBoolean connectionIsClosed = new AtomicBoolean(false);
    private final ConnectionManagerConfiguration configuration;
    private final ExecutorService sharedExecutor = Executors.newSingleThreadExecutor(new ThreadFactoryBuilder()
            .setNameFormat("rabbitmq-shared-pool-%d")
            .build());
    private final HealthMetrics metrics = new HealthMetrics(this);

    public ConsumeConnectionManager(@NotNull ConnectionManagerConfiguration connectionManagerConfiguration,
            Runnable onFailedRecoveryConnection, ConnectionFactory factory) {
        this.configuration = Objects.requireNonNull(connectionManagerConfiguration, "Connection manager configuration can not be null");
        factory.setExceptionHandler(new ExceptionHandler() {
            @Override
            public void handleUnexpectedConnectionDriverException(Connection conn, Throwable exception) {
                turnOffReadiness(exception);
            }

            @Override
            public void handleReturnListenerException(Channel channel, Throwable exception) {
                turnOffReadiness(exception);
            }

            @Override
            public void handleConfirmListenerException(Channel channel, Throwable exception) {
                turnOffReadiness(exception);
            }

            @Override
            public void handleBlockedListenerException(Connection connection, Throwable exception) {
                turnOffReadiness(exception);
            }

            @Override
            public void handleConsumerException(Channel channel, Throwable exception, Consumer consumer, String consumerTag, String methodName) {
                turnOffReadiness(exception);
            }

            @Override
            public void handleConnectionRecoveryException(Connection conn, Throwable exception) {
                turnOffReadiness(exception);
            }

            @Override
            public void handleChannelRecoveryException(Channel ch, Throwable exception) {
                turnOffReadiness(exception);
            }

            @Override
            public void handleTopologyRecoveryException(Connection conn, Channel ch, TopologyRecoveryException exception) {
                turnOffReadiness(exception);
            }

            private void turnOffReadiness(Throwable exception) {
                metrics.getReadinessMonitor().disable();
                LOGGER.debug("Set RabbitMQ readiness to false. RabbitMQ error", exception);
            }
        });

        factory.setConnectionRecoveryTriggeringCondition(shutdownSignal -> !connectionIsClosed.get() && getConnectionRecoveryTriggeringCondition(onFailedRecoveryConnection));

        factory.setRecoveryDelayHandler(recoveryAttempts -> getRecoveryDelay());

        factory.setSharedExecutor(sharedExecutor);

        try {
            connection = factory.newConnection();
            metrics.getReadinessMonitor().enable();
            LOGGER.debug("Set RabbitMQ readiness to true");
        } catch (IOException | TimeoutException e) {
            metrics.getReadinessMonitor().disable();
            LOGGER.debug("Set RabbitMQ readiness to false. Can not create consume connection", e);
            throw new IllegalStateException("Failed to create RabbitMQ consume connection using configuration", e);
        }

        attachBlockedListener();
        attachRecoveryListener();
    }

    private boolean getConnectionRecoveryTriggeringCondition(Runnable onFailedRecoveryConnection) {
        int tmpCountTriesToRecovery = connectionRecoveryAttempts.get();

        if (tmpCountTriesToRecovery < configuration.getMaxRecoveryAttempts()) {
            LOGGER.info("Try to recovery consume connection to RabbitMQ. Count tries = {}", tmpCountTriesToRecovery + 1);
            return true;
        }
        LOGGER.error("Can not connect to RabbitMQ. Count tries = {}", tmpCountTriesToRecovery);
        if (onFailedRecoveryConnection == null) {
            throw new IllegalStateException("Cannot recover connection to RabbitMQ");
        }
        onFailedRecoveryConnection.run();
        return false;
    }

    private int getRecoveryDelay() {
        int tmpCountTriesToRecovery = connectionRecoveryAttempts.getAndIncrement();

        int recoveryDelay = configuration.getMinConnectionRecoveryTimeout()
                + (configuration.getMaxRecoveryAttempts() > 1
                ? (configuration.getMaxConnectionRecoveryTimeout() - configuration.getMinConnectionRecoveryTimeout())
                / (configuration.getMaxRecoveryAttempts() - 1)
                * tmpCountTriesToRecovery
                : 0);

        LOGGER.info("Recovery delay for '{}' try = {}", tmpCountTriesToRecovery, recoveryDelay);
        return recoveryDelay;
    }

    private void attachBlockedListener() {
        connection.addBlockedListener(new BlockedListener() {
            @Override
            public void handleBlocked(String reason) {
                LOGGER.warn("RabbitMQ blocked consume connection: {}", reason);
            }

            @Override
            public void handleUnblocked() {
                LOGGER.warn("RabbitMQ unblocked consume connection");
            }
        });
    }

    private void attachRecoveryListener() {
        if (connection instanceof Recoverable) {
            Recoverable recoverableConnection = (Recoverable)connection;
            recoverableConnection.addRecoveryListener(getRecoveryListener());
            LOGGER.debug("Recovery listener was added to consume connection.");
        } else {
            throw new IllegalStateException("Consume connection does not implement Recoverable. Can not add RecoveryListener to it");
        }
    }

    private RecoveryListener getRecoveryListener() {
        return new RecoveryListener() {
            @Override
            public void handleRecovery(Recoverable recoverable) {
                LOGGER.debug("Count tries to recovery consume connection reset to 0");
                connectionRecoveryAttempts.set(0);
                metrics.getReadinessMonitor().enable();
                LOGGER.debug("Set RabbitMQ readiness to true");
            }

            @Override
            public void handleRecoveryStarted(Recoverable recoverable) {
            }
        };
    }

    public void shutdownSharedExecutor(int closeTimeout) {
        sharedExecutor.shutdown();
        try {
            if (!sharedExecutor.awaitTermination(closeTimeout, TimeUnit.MILLISECONDS)) {
                LOGGER.error("Executor is not terminated during {} millis", closeTimeout);
                List<Runnable> runnables = sharedExecutor.shutdownNow();
                LOGGER.error("{} task(s) was(were) not finished", runnables.size());
            }
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }

    public void closeConnection(int closeTimeout) {
        if (connectionIsClosed.getAndSet(true)) {
            return;
        }
        if (connection.isOpen()) {
            try {
                // We close the connection and don't close channels
                // because when a channel's connection is closed, so is the channel
                connection.close(closeTimeout);
            } catch (IOException e) {
                LOGGER.error("Cannot close consume connection", e);
            }
        }
    }

    public Connection getConnection() {
        return connection;
    }
}

File: src/main/kotlin/com/exactpro/th2/common/schema/message/impl/rabbitmq/connection/ConsumeConnectionManager.kt
/*
 * Copyright 2024 Exactpro (Exactpro Systems Limited)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

package com.exactpro.th2.common.schema.message.impl.rabbitmq.connection

import com.exactpro.th2.common.schema.message.DeliveryMetadata
import com.exactpro.th2.common.schema.message.ExclusiveSubscriberMonitor
import com.exactpro.th2.common.schema.message.ManualAckDeliveryCallback
import com.exactpro.th2.common.schema.message.ManualAckDeliveryCallback.Confirmation
import com.exactpro.th2.common.schema.message.impl.OnlyOnceConfirmation.Companion.wrap
import com.exactpro.th2.common.schema.message.impl.rabbitmq.configuration.ConnectionManagerConfiguration
import com.exactpro.th2.common.schema.message.impl.rabbitmq.configuration.RabbitMQConfiguration
import com.rabbitmq.client.Channel
import com.rabbitmq.client.Delivery
import com.rabbitmq.client.CancelCallback
import com.rabbitmq.client.BlockedListener
import com.rabbitmq.client.ShutdownSignalException

import mu.KotlinLogging
import java.io.IOException
import java.util.concurrent.TimeUnit
import java.util.concurrent.TimeoutException
import java.util.concurrent.atomic.AtomicInteger

class ConsumeConnectionManager(
    connectionName: String,
    rabbitMQConfiguration: RabbitMQConfiguration,
    connectionManagerConfiguration: ConnectionManagerConfiguration
) : ConnectionManager(connectionName, rabbitMQConfiguration, connectionManagerConfiguration) {
    private val subscriberName = if (connectionManagerConfiguration.subscriberName.isNullOrBlank()) {
        (DEFAULT_SUBSCRIBER_NAME_PREFIX + System.currentTimeMillis()).also {
            LOGGER.info { "Subscribers will use the default name: $it" }
        }
    } else {
        connectionManagerConfiguration.subscriberName + "." + System.currentTimeMillis()
    }
    private val nextSubscriberId: AtomicInteger = AtomicInteger(1)

    override fun createBlockedListener(): BlockedListener = object : BlockedListener {
        override fun handleBlocked(reason: String) {
            LOGGER.info { "RabbitMQ blocked consume connection: $reason" }
        }

        override fun handleUnblocked() {
            LOGGER.info("RabbitMQ unblocked consume connection")
        }
    }

    @Throws(IOException::class)
    fun queueDeclare(): String {
        val holder = ChannelHolder(this::createChannel, this::waitForConnectionRecovery, configurationToOptions())
        return holder.mapWithLock { channel ->
            channel.queueDeclare(
                "",  // queue name
                false,  // durable
                true,  // exclusive
                false,  // autoDelete
                emptyMap()
            ).queue.also { queue ->
                LOGGER.info { "Declared exclusive '$queue' queue" }
                putChannelFor(PinId.forQueue(queue), holder)
            }
        }
    }

    @Throws(IOException::class, TimeoutException::class)
    fun queueExclusiveDeclareAndBind(exchange: String): String {
        createChannel().use { channel ->
            val queue = channel.queueDeclare().queue
            channel.queueBind(queue, exchange, EMPTY_ROUTING_KEY)
            LOGGER.info { "Declared the '$queue' queue to listen to the '$exchange'" }
            return queue
        }
    }

    @Throws(IOException::class, InterruptedException::class)
    fun basicConsume(
        queue: String,
        deliverCallback: ManualAckDeliveryCallback,
        cancelCallback: CancelCallback
    ): ExclusiveSubscriberMonitor {
        val pinId = PinId.forQueue(queue)
        val holder = getOrCreateChannelFor(pinId, SubscriptionCallbacks(deliverCallback, cancelCallback))
        val tag = holder.retryingConsumeWithLock({ channel: Channel ->
            channel.basicConsume(
                queue, false, subscriberName + "_" + nextSubscriberId.getAndIncrement(),
                { tagTmp: String, delivery: Delivery ->
                    try {
                        val envelope = delivery.envelope
                        val deliveryTag = envelope.deliveryTag
                        val routingKey = envelope.routingKey
                        LOGGER.trace { "Received delivery $deliveryTag from queue=$queue routing_key=$routingKey" }

                        val wrappedConfirmation: Confirmation = object : Confirmation {
                            @Throws(IOException::class)
                            override fun reject() {
                                holder.withLock { ch: Channel ->
                                    try {
                                        ch.basicReject(deliveryTag, false)
                                    } catch (e: IOException) {
                                        LOGGER.warn { "Error during basicReject of message with deliveryTag = $deliveryTag inside channel #${ch.channelNumber}: $e" }
                                        throw e
                                    } catch (e: ShutdownSignalException) {
                                        LOGGER.warn { "Error during basicReject of message with deliveryTag = $deliveryTag inside channel #${ch.channelNumber}: $e" }
                                        throw e
                                    } finally {
                                        holder.release { metrics.readinessMonitor.enable() }
                                    }
                                }
                            }

                            @Throws(IOException::class)
                            override fun confirm() {
                                holder.withLock { ch: Channel ->
                                    try {
                                        // because delivery tags are scoped per channel,
                                        // deliveries must be acknowledged on the same channel they were received on.
                                        ch.basicAck(deliveryTag, false)
                                    } catch (e: Exception) {
                                        LOGGER.warn { "Error during basicAck of message with deliveryTag = $deliveryTag inside channel #${ch.channelNumber}: $e" }
                                        throw e
                                    } finally {
                                        holder.release { metrics.readinessMonitor.enable() }
                                    }
                                }
                            }
                        }

                        val confirmation = wrap("from $routingKey to $queue", wrappedConfirmation)

                        holder.withLock(Runnable {
                            holder.acquireAndSubmitCheck {
                                channelChecker.schedule<Boolean>(
                                    {
                                        holder.withLock(Runnable {
                                            LOGGER.warn { "The confirmation for delivery $deliveryTag in queue=$queue routing_key=$routingKey was not invoked within the specified delay" }
                                            if (holder.reachedPendingLimit()) {
                                                metrics.readinessMonitor.disable()
                                            }
                                        })
                                        false // to cast to Callable
                                    },
                                    configuration.confirmationTimeout.toMillis(),
                                    TimeUnit.MILLISECONDS
                                )
                            }
                        })
                        val redeliver = envelope.isRedeliver
                        deliverCallback.handle(DeliveryMetadata(tagTmp, redeliver), delivery, confirmation)
                    } catch (e: Exception) {
                        LOGGER.error("Cannot handle delivery for tag $tagTmp: ${e.message}", e)
                    } catch (e: RuntimeException) {
                        LOGGER.error("Cannot handle delivery for tag $tagTmp: ${e.message}", e)
                    }
                }, cancelCallback
            )
        }, configuration)

        return RabbitMqSubscriberMonitor(holder, queue, tag) { channel, consumerTag -> channel.basicCancel(consumerTag) }
    }

    override fun recoverSubscriptionsOfChannel(pinId: PinId, channel: Channel, holder: ChannelHolder) {
        channelChecker.execute {
            holder.withLock(Runnable {
                try {
                    val subscriptionCallbacks = holder.getCallbacksForRecovery(channel)

                    if (subscriptionCallbacks != null) {
                        LOGGER.info { "Changing channel for holder with pin id: $pinId" }

                        val removedHolder = channelsByPin.remove(pinId)
                        check(removedHolder === holder) { "Channel holder has been replaced" }

                        basicConsume(
                            pinId.queue,
                            subscriptionCallbacks.deliverCallback,
                            subscriptionCallbacks.cancelCallback
                        )
                    }
                } catch (e: InterruptedException) {
                    Thread.currentThread().interrupt()
                    LOGGER.info("Recovering channel's subscriptions interrupted", e)
                } catch (e: Throwable) {
                    // this code executed in executor service and exception thrown here will not be handled anywhere
                    LOGGER.error("Failed to recovery channel's subscriptions", e)
                }
            })
        }
    }

    private class RabbitMqSubscriberMonitor(
        private val holder: ChannelHolder,
        override val queue: String,
        private val tag: String,
        private val action: CancelAction
    ) : ExclusiveSubscriberMonitor {
        @Throws(IOException::class)
        override fun unsubscribe() {
            holder.unsubscribeWithLock(tag, action)
        }
    }

    companion object {
        private val LOGGER = KotlinLogging.logger {}
        private const val DEFAULT_SUBSCRIBER_NAME_PREFIX = "rabbit_mq_subscriber."
    }
}